### PR TITLE
Enable voting on projects

### PR DESCRIPTION
**Pull request by daniel-patterson-ezcater**
## What did we change?

Added the ability for a user to vote on all the awards at a particular event, voting for one award at a time.

This keeps the state (who's voting, where they are in the list of awards, whether they've finished voting) in URL params. This isn't particularly reliable - users could easily get around this and vote multiple times - but we can't really deal with that well until we have a `User` model and authentication.

## Why are we doing this?

So people can vote on which projects deserve awards at our current hackathon.

## What isn't included?

* Linking to voting from anywhere (we need to decide where we want this to be from)
* Editing votes (I'll probably add this next)

## Checklist
- [x] Automated Tests (check all that apply)
  - [ ] Unit
  - [x] Request/Integration
  - [ ] Feature/System
- [ ] Migrations Reviewed (Zero Downtime https://ezcater.atlassian.net/l/c/F1u4298i)

### CODE REVIEW

File: app/controllers/votes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  private
+
+  def load_event
+    @event = Event.find(params[:event_id])
+  end
+
+  def load_control_strategy
+    @control_strategy = ::Voting::UrlControlledVotingStrategy.new(params: params)
+  end
```
Comment (daniel-patterson-ezcater) [inline app/controllers/votes_controller.rb:30]: This strategy object is meant to be the only bit that is coupled to using the URL params as the place we store state and deciding where the user is in the voting flow. 

My idea is we could replace this with storing state in a cookie to be a _little_ less likely for people to cheat, or we could replace it with something that looks up the user in the database once we have authentication. 

This is based on a suggestion from @squidpunch that I really liked.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  def load_state
+    @state = ::Voting::State.new(control_strategy: @control_strategy, event: @event)
+  end
```
Comment (daniel-patterson-ezcater) [inline app/controllers/votes_controller.rb:34]: This object is agnostic of where the user's state is stored (it pulls that from the strategy), but is kind of a presenter object, exposing the information needed to build the voting interface. 

Ideally this would be tested as it's easier to test edge cases on this than in views / integration tests. However, I opted to just have high-level integration tests as this is a hackathon ...

File: app/models/event.rb [hunk 3]
```diff
@@ -5,4 +5,8 @@ class Event < ApplicationRecord
 
   has_many :projects, dependent: :restrict_with_exception
   has_many :votes, dependent: :destroy
+
+  def voting_enabled?
+    true # Perhaps change this based on a status / stage column?
+  end
```
Comment (daniel-patterson-ezcater) [inline app/models/event.rb:11]: Just an idea - I'm happy to remove this

File: app/controllers/votes_controller.rb (hunk 1 above)
Comment (tristandunn) [reply app/controllers/votes_controller.rb:30]: If we store the name in the session or a cookie, then the `Vote` model validation and unique index will prevent multiple votes by the same person for the same event and award. And I figured we'd swap the name to a user ID once we have authentication.

File: app/models/event.rb (hunk 3 above)
Comment (tristandunn) [reply app/models/event.rb:11]: Think it's fine to leave for now but we should discuss how to handle it. We need no voting, voting, and voting finalized states so far I believe.

File: app/services/voting/state.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      return @_next_award if defined? @_next_award
+
+      @_next_award = if current_award.nil?
+                       awards.first
+                     elsif current_award == awards.last
+                       nil
+                     else
+                       awards[current_award_index + 1]
+                     end
+    end
```
Comment (tristandunn) [inline app/services/voting/state.rb:46]: 🎱 You might be able to do this in the database by excluding awards already voted on for by the user. Also might be overkill for now.

File: app/services/voting/state.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    def awards
+      @_awards ||= Award.order(:title)
+    end
+
+    def current_award
+      return @_current_award if defined? @_current_award
+
+      @_current_award = current_award_id.nil? ? nil : Award.find(current_award_id)
```
Comment (tristandunn) [inline app/services/voting/state.rb:33]: 🎱 
```suggestion
      @_current_award = Award.find_by(id: current_award_id)
```

File: app/views/votes/_voting.html.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<h2>"<%= state.current_award.title %>"</h2>
+<p>Award <%= state.current_award_index + 1 %> of <%= state.awards.size %></p>
+
+<%= form_with(url: "/events/#{state.event.id}/vote", local: true, method: :post) do |f| %>
```
Comment (tristandunn) [inline app/views/votes/_voting.html.erb:4]: 🎱 Can the URL be a route helper?

File: app/controllers/votes_controller.rb (hunk 1 above)
Comment (daniel-patterson-ezcater) [reply app/controllers/votes_controller.rb:30]: This is essentially doing that, but storing the name in a url parameter. We could swap that to a cookie easily enough, but I'd prefer to do that as a follow-on PR so we definitely have something working.

I agree we'd swap to a user id eventually.

File: app/models/event.rb (hunk 3 above)
Comment (daniel-patterson-ezcater) [reply app/models/event.rb:11]: I'm going to have a shot at adding a `voting_status` column in a follow-on PR to add those three states.

File: app/services/voting/state.rb (hunk 5 above)
Comment (daniel-patterson-ezcater) [reply app/services/voting/state.rb:33]: I kept `find` here as I wanted this to raise if the award id is invalid. Our error handling isn't great, but we seem to already use `find` a lot as a a rudimentary form of validation.

File: app/services/voting/state.rb (hunk 4 above)
Comment (daniel-patterson-ezcater) [reply app/services/voting/state.rb:46]: I had a shot, but decided against it as this method makes it a little easier to implement allowing someone to go back and edit their votes while voting is still open.

File: app/services/voting/state.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    end
+
+    delegate :username,
+             :user_started_voting?,
+             :user_finished_voting?,
+             :current_award_id,
+             to: :@control_strategy
+
+    def ordered_projects
+      @_ordered_projects ||= projects.order("ideas.name")
```
Comment (JessSumner) [inline app/services/voting/state.rb:19]: With the delegation of `name` to `idea` would the following work ?
```suggestion
      @_ordered_projects ||= projects.order("name")
```

Comment (daniel-patterson-ezcater) [reply app/services/voting/state.rb:19]: Apparently not. I just tried it and got:

```
Error:
       PG::UndefinedColumn: ERROR:  column "name" does not exist
       LINE 1: ...OM "projects" WHERE "projects"."event_id" = $1 ORDER BY name                    
```

I guess the delegation only works in Ruby and isn't propagated by `ActiveRecord` into SQL.

File: spec/features/user_votes_on_awards_spec.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      "Most likely to break production",
+      "Most likely to make it into production",
+    ].map { |title| create(:award, title: title) }
+  end
+
+  let!(:projects) do
+    [
+      "A useful idea",
+      "An idea that just might work",
+      "Hey, how about we convert the entire codebase to COBOL?",
```
Comment (JessSumner) [inline spec/features/user_votes_on_awards_spec.rb:19]: Next hackathon!!! 😆

File: spec/features/user_votes_on_awards_spec.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  scenario "user votes for all awards" do
+    original_vote_count = Vote.count
+
+    visit new_event_vote_path(event)
+    fill_in :username, with: "Test User"
+    click_button "start_voting"
+
+    expect(page).to have_content "Most likely to break production"
+    expect(page).to have_content "Award 1 of 2"
+    choose "Hey, how about we convert the entire codebase to COBOL?"
```
Comment (JessSumner) [inline spec/features/user_votes_on_awards_spec.rb:61]: A worthy winner!

**Review by JessSumner — APPROVED** [review]

File: app/services/voting/state.rb (hunk 7 above)
Comment (JessSumner) [reply app/services/voting/state.rb:19]: Oh good to know. Thank you for testing!
